src/main/scala/org/broadinstitute/dsde/workbench/sam/azure/AzureService.scala

```diff
@@ -126,6 +133,70 @@ class AzureService(crlService: CrlService, directoryDAO: DirectoryDAO, azureMana
       createdPet <- directoryDAO.createPetManagedIdentity(petToCreate, samRequestContext)
     } yield (createdPet, true)
 
+  def getOrCreateActionManagedIdentity(
+      resource: FullyQualifiedResourceId,
+      resourceAction: ResourceAction,
+      mrgCoordinates: ManagedResourceGroupCoordinates,
+      samUser: SamUser,
+      samRequestContext: SamRequestContext
+  ): IO[(ActionManagedIdentity, Boolean)] = {
+    val id = ActionManagedIdentityId(resource, resourceAction, mrgCoordinates)
+    for {
+      hasPermission <- policyEvaluatorService.hasPermission(resource, resourceAction, samUser.id, samRequestContext)
+      _ <- IO.raiseWhen(!hasPermission)(
+        new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, s"User ${samUser.id} does not have $resourceAction on $resource"))
+      )
+      existingAmiOpt <- directoryDAO.loadActionManagedIdentity(id, samRequestContext)
+      ami <- existingAmiOpt match {
+        // pet exists in Sam DB - return it
+        case Some(p) => IO.pure((p, false))
+        // pet does not exist in Sam DB - create it
+        case None => createActionManagedIdentity(id, samRequestContext)
+      }
+    } yield ami
+
+  }
+
+  private def createActionManagedIdentity(
```

Comment:
Is there anything in call tree that validates that the MRG is trusted like by calling `validateManagedResourceGroup` or ensuring that the MRG coordinates exist in the db?

Reply:
Done